{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "    \n",
    "    Analyses of ribosome profiling data related to the HrpA project\n",
    "    \n",
    "    Copyright (C) 2021  Allen Buskirk, buskirk@jhmi.edu\n",
    "    \n",
    "    This program is free software: you can redistribute it and/or modify\n",
    "    it under the terms of the GNU General Public License as published by\n",
    "    the Free Software Foundation, either version 3 of the License, or\n",
    "    (at your option) any later version.\n",
    "    This program is distributed in the hope that it will be useful,\n",
    "    but WITHOUT ANY WARRANTY; without even the implied warranty of\n",
    "    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n",
    "    GNU General Public License for more details.\n",
    "    You should have received a copy of the GNU General Public License\n",
    "    along with this program.  If not, see <http://www.gnu.org/licenses/>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "## this notebook is for analysis of samples ac21-26\n",
    "## mapping reads to the second version of the MG1655 genome\n",
    "## all samples express nanoluc-IRAGP reporter, analysis for which are in hrpA_reporter notebook, elsewhere\n",
    "\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "import os\n",
    "from BCBio import GFF\n",
    "import ribo_utilities\n",
    "\n",
    "FP_names = [\n",
    "    'ac21', 'ac22', 'ac23', 'ac24', 'ac25', 'ac26'\n",
    "    ]\n",
    "\n",
    "inpath      = '/home/annie/profiling_data/hrpA_lowERY_rep1_data/'    \n",
    "path_script = '/home/annie/code/profiling/code_hrpA/'\n",
    "\n",
    "paths_in = {\n",
    "    'path_fastq'     : inpath  + 'fastq/', \n",
    "    'path_gff'       : path_script + 'coli2.gff',\n",
    "    'path_bowtie'    : 'bowtie',\n",
    "    'index_ladder'   : path_script + 'indexes/ladder/ladder',\n",
    "    'index_trna'     : path_script + 'indexes/MG1655/MG1655_tRNA',\n",
    "    'index_rrna'     : path_script + 'indexes/MG1655/MG1655_rRNA',\n",
    "    'index_chr'      : path_script + 'indexes/MG1655/MG1655_genome',\n",
    "    }\n",
    "\n",
    "paths_out = {\n",
    "    'path_filter'       : inpath  + 'filterdata/',\n",
    "    'path_ladder'       : inpath  + 'alignments/ladder/',\n",
    "    'path_trna'         : inpath  + 'alignments/tRNA/',\n",
    "    'path_rrna'         : inpath  + 'alignments/rRNA/',\n",
    "    'path_chr'          : inpath  + 'alignments/chr/',\n",
    "    'path_temp'         : inpath  + 'tmpds/',\n",
    "    'path_density'      : inpath  + 'density/',\n",
    "    'path_wig'          : inpath  + 'wigfiles/',\n",
    "    'path_log'          : inpath  + 'log/',\n",
    "    'path_genelists'    : inpath  + 'analysis/genelists/',\n",
    "    'path_pauses'       : inpath  + 'analysis/pauses/'\n",
    "    }\n",
    "\n",
    "for i in paths_out:\n",
    "    if not os.path.exists(paths_out[i]):\n",
    "        os.makedirs(paths_out[i])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## use cutadapt 2.1 version to remove the linker from the reads in the FASTQ files \n",
    "## and the UMI 4 nt before and 6 nt after the footprint itself\n",
    "\n",
    "\n",
    "import os\n",
    "\n",
    "minlength = 10\n",
    "\n",
    "linker = 'NNNNNNCACTCGGGCACCAAGGAC'\n",
    "\n",
    "files = FP_names\n",
    "\n",
    "#print \"log saved at paths_out['path_filter']\"\n",
    "# teach cutadapt to save output or save by hand from the notebook terminal\n",
    "\n",
    "for fname in files:\n",
    "    file_in = paths_in['path_fastq'] + fname + '.fastq'\n",
    "    file_out = paths_out['path_filter'] + fname + '_filtered_trimmed.fastq'\n",
    "\n",
    "    cutadapt_command = 'cutadapt -a %s --discard-untrimmed %s | cutadapt -u 4 -m %d - > %s' % (\n",
    "        linker, \n",
    "        file_in, \n",
    "        minlength,\n",
    "        file_out\n",
    "        )\n",
    "\n",
    "    os.system(cutadapt_command)\n",
    "    print fname + \" completed\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# alignments using bowtie 1.2.2\n",
    "# this function aligns to a ladder index, then the tRNA index, then the rRNA index, then the chromosome\n",
    "# tRNA and rRNA match alignments stored in the tmpds folder\n",
    "# makes tmpds folder and alignment folders automatically\n",
    "\n",
    "import os\n",
    "import sys\n",
    "\n",
    "files = FP_names\n",
    "\n",
    "######################################################\n",
    "\n",
    "path_l = paths_out['path_ladder']\n",
    "path_t = paths_out['path_trna']\n",
    "path_r = paths_out['path_rrna']\n",
    "path_chr = paths_out['path_chr']\n",
    "path_temp = paths_out['path_temp']\n",
    "\n",
    "if os.path.exists(paths_out['path_filter']+files[0]+'_filtered_trimmed.fastq'):\n",
    "    trimmed = '_filtered_trimmed.fastq '\n",
    "elif os.path.exists(paths_out['path_filter']+files[0]+'-trimmed.fastq'):\n",
    "    trimmed = '-trimmed.fastq '\n",
    "else:\n",
    "    sys.exit(\"error: trimmed file name\")\n",
    "    \n",
    "for fname in files:\n",
    "    print fname\n",
    "    \n",
    "    #first, align to ladder index to subtract any ladder oligos that were cloned\n",
    "    bowtie_l = paths_in['path_bowtie'] + ' -v 2 -y -m 1 -a --best --strata -S -p 2 --un '\n",
    "    bowtie_l += path_l + fname + '_nomatch.fastq --max ' + path_l + fname + '_multi.fastq --al '\n",
    "    bowtie_l += path_l + fname + '_match.fastq ' + paths_in['index_ladder'] + ' '\n",
    "    bowtie_l += paths_out['path_filter'] + fname + trimmed + ' 1> ' + path_temp + fname + '_ladder_match.SAM'\n",
    "    bowtie_l += ' 2> ' + paths_out['path_log'] + fname\n",
    "    os.system(bowtie_l)\n",
    "    os.system(\"echo 'End ladder alignment' >> \" + paths_out['path_log'] + fname)\n",
    "\n",
    "    # second, align to tRNA index\n",
    "    bowtie_tRNA = paths_in['path_bowtie'] + ' -v 2 -y -m 1 -a --best --strata -S -p 2 --un '\n",
    "    bowtie_tRNA += path_t + fname + '_nomatch.fastq --max ' + path_t + fname + '_multi.fastq --al '\n",
    "    bowtie_tRNA += path_t + fname + '_match.fastq ' + paths_in['index_trna'] + ' '\n",
    "    bowtie_tRNA += path_l + fname + '_nomatch.fastq ' + ' 1> ' + path_temp + fname + '_tRNA_match.SAM'\n",
    "    bowtie_tRNA += ' 2>> ' + paths_out['path_log'] + fname \n",
    "    os.system(bowtie_tRNA)\n",
    "    os.system(\"echo 'End tRNA alignment' >> \" + paths_out['path_log'] + fname)\n",
    "\n",
    "    # third, align to the rRNA index\n",
    "    bowtie_rRNA = paths_in['path_bowtie'] + ' -v 2 -y -m 1 -a --best --strata -S -p 2 --un '\n",
    "    bowtie_rRNA += path_r + fname + '_nomatch.fastq --max ' + path_r + fname + '_multi.fastq --al '\n",
    "    bowtie_rRNA += path_r + fname + '_match.fastq ' + paths_in['index_rrna'] + ' '\n",
    "    bowtie_rRNA += path_t + fname + '_nomatch.fastq ' + ' 1> ' + path_temp + fname + '_rRNA_match.SAM'\n",
    "    bowtie_rRNA += ' 2>> ' + paths_out['path_log'] + fname \n",
    "    os.system(bowtie_rRNA)\n",
    "    os.system(\"echo 'End rRNA alignment' >> \" + paths_out['path_log'] + fname)\n",
    "\n",
    "    # finally align to the chr index\n",
    "    bowtie_chr = paths_in['path_bowtie'] + ' -v 2 -y -m 1 -a --best --strata -S -p 2 --un '\n",
    "    bowtie_chr += path_chr + fname + '_nomatch.fastq --max ' + path_chr + fname + '_multi.fastq --al '\n",
    "    bowtie_chr += path_chr + fname + '_match.fastq ' + paths_in['index_chr'] + ' '\n",
    "    bowtie_chr += path_r + fname + '_nomatch.fastq ' + ' 1> ' + path_chr + fname + '_match.SAM'\n",
    "    bowtie_chr += ' 2>> ' + paths_out['path_log'] + fname \n",
    "    os.system(bowtie_chr)\n",
    "    os.system(\"echo 'End chromosome alignment' >> \" + paths_out['path_log'] + fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### creating ribosome density files, saving pickle file and WIG file\n",
    "### by hand, you must collect all the output and save in the reads.csv file in the density folder with format:\n",
    "### filename,mapped_reads\n",
    "\n",
    "files = FP_names\n",
    "minlength = 10\n",
    "maxlength = 35\n",
    "threads = 4\n",
    "\n",
    "ribo_utilities.run_density(files, minlength, maxlength, threads, paths_in, paths_out, 'ribo_seq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
